### ISSUE TITLE
Word wrap should not be disabled when accessibility is turned on

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#95428]
We disable word wrap when accessibility is turned on. I would remove this limiation due to three reasons
1) User feedback https://twitter.com/LeonieWatson/status/1250717463672688647
2) Now the status bar is accessibile and cltr + g announces line number so it is much easier to figure out what line you are on
3) WordWrap is off by default if the user turns it on he or she should be aware of it

**Comment by isidorn** [comment]
After discussing with @alexdima he informed me that we disable word wrap due to this issue https://github.com/Microsoft/vscode/issues/27766
I justr tried and it is still an case, if we would to enable word wrap when screen reader attached Screen Readers (orca, nvda) would read the entire long line when user navigates to its wrapped part.

As @alexdima puts it we we first need to decide what content do we want to put in the `textarea`.
At the moment we have put the model content and not the view line content we might want to put the view line content, which would make it that a line like `"a long line"` might be put as `"a \nlong \nline"`, which means we put non-existing `"\n"` in there. The problem has always been how to communicate a wrapping point. Using `"\n"` for a wrapping point feels incorrect for me, because screen reader will read a new line, but in reality it is not a new line.

@joanmarie @leonardder @marcozehe @jvesouza @marlon-sousa do you have some recommendation on how should we communicate a wrapping point to the users?

In case somebody wants to try word wrapping with screen readers, do the following:
1. Remove this [if check](https://github.com/microsoft/vscode/blob/cb0667f934c1c142bdbd6dd30619ce0233c8ff2d/src/vs/editor/common/config/editorOptions.ts#L3240)
2. [Build vscode out of source](https://github.com/microsoft/vscode/wiki/How-to-Contribute#build-and-run)
3. Set the following setting 	`"editor.wordWrap": "on"`,
4. Have a very long line which gets wrapped and check what screen reader reads when you navigate up / down

**Comment by MarcoZehe** [comment]
Honestly, I think the fact that WordWrap is disabled by default if I use a screen reader, is one of the greatest features of Code. As a braille user, especially when dealing with real code, not text like markdown or HTmL, I would hate it if I cursor down and land in the middle of a line which on the braille display appears as a new line at the beginning of the braille output. Imagine a Python file where indentation is of the essence syntactically, and then getting lines jumping back and forth on the braille display as you cursor up and down. Having to constantly turn word wrap on or off would make dealing with different code parts a real chore for me. And remembering to turn it off or such, too.

Also for speech, when dealing with code, I find it much harder to deal with wrapped lines rather than a long line, which when reviewing, I'd move through either word by word, or character by character, anyway.

I am currently not set up to build VS Code, so someone else would have to do the testing whether cursoring up and down through a wrapped line would still read everything, not just the part that the cursor lands on, the "virtual line", so to speak. But personally, I found it one of the greatest first-time user experiences that WordWrap was turned off for me. No ambiguity, no fuzz.

**Comment by isidorn** [comment]
@marcozehe thanks a lot for feedback, very useful to know this. I tried out latest vscode with screen readers and as I mentioned above  NVDA and Orca would read the entire long line when user navigates to its wrapped part. So it does not behave well.

**Comment by leonardder** [comment]
I agree with @MarcoZehe that word wrap in a code editor usually gives a bad experience with screen readers.

It just occurred to me that somehow, this issue is very similar too the issue with broken word navigation in that Code isn't able to override the line offsets that are exposed by IAccessible2.

Having said that, it might help to have a bit of a technical description about what happens exactly when word wrap is on. Is it only a change to how things are displayed and how up/down arrow work? On github for example, there's also word wrap in the comment field, and that works just fine. I wonder how VS Code differs from this default behavior that's probably triggered by changing the width of the area.

**Comment by wsmelton** [comment]
If this is to stay in place please consider providing some warning or message when users attempt to "enable" word wrap...or sit there for an hour figuring out why lines are not wrapping 😖 

I've also determined with Windows 10, version 2004 there is a feature that lets you know customize the text cursor...this apparently is being considered as a screen reader in VS Code so it disabled word wrap in my editor.

**Comment by isidorn** [comment]
@wsmelton agree, if the user enables word wrap and it is disabled we should do a warning via the notification service. We are open to a PR that fixes this.

**Comment by alexdima** [comment]
@wsmelton A notification is shown when a Screen Reader is detected. The notification reads: "Are you using a screen reader to operate VS Code? (Certain features like word wrap are disabled when using a screen reader)".

**Comment by wsmelton** [comment]
There was no indication it was going to disable word wrap for me with insiders build.

**Comment by alexdima** [comment]
@wsmelton Did you by any chance configure `"editor.accessibilitySupport"` in user settings? If the setting is left to be `auto` (the default), then VS Code will prompt with a notification when it detects a Screen Reader. The question clearly states that word wrap will be disabled and then you can answer Yes or No. If you answer Yes the setting `"editor.accessibilitySupport"` is written to user settings to be `"on"`. If you answer No the setting is written to be `"off"`.

**Comment by wsmelton** [comment]
It is set to Auto, and I've enabled and disabled an Ease of Use settings in Windows 10...restarted Insiders a few times but never seem to get a prompt.

**Comment by jhomme** [comment]
Hi,
As a user, I would want to choose whether to have it on or off. For code, I agree what Marco says. For writing in Markdown, if using NVDA, which cannot navigate by sentences in most places, editing can be slow. When using JAWS, which can read by sentences in all places that I am aware of, this is not an issue in VS Code. Also, in my experience so far, with NVDA, the longer the lines are, the more slugginsh it is.


> 
> 
e> @MarcoZehe thanks a lot for feedback, very useful to know this. I tried out latest vscode with screen readers and as I mentioned above NVDA and Orca would read the entire long line when user navigates to its wrapped part. So it does not behave well. I just left the instructions for bulding in case VoiceOver owners might want to experiment.

**Comment by prater-devin** [comment]
I'm mainly, for now, using VS Code for writing Markdown, due to its fabulous Language Tool extension which works much better than with Emacs with Emacspeak. Emacspeak gets around wrapping by playing a sound when the next logical line is reached, rather than a visual line. I hope word wrap can be done, at least in text and Markdown modes, because VS Code works great otherwise.

**Comment by jhomme** [comment]
Hi Devin,

Get the Rewrap extension.

 

Jim H

 

==========

Jim Homme

Skype: jim.homme

FreeChess: jhomme

Twitter: jimhomme <https://twitter.com/jimhome> 

Facebook: jimhomme

Website: jimhomme.com <https://www.jimhomme.com/>

**Comment by prater-devin** [comment]
I have that, but having to press Alt + Q on every paragraph, and when I’m done typing, is a lot of work for a workaround. Also, when other people view my work on Github, sometimes the line breaks actually make new paragraphs. Although, that may just be an Emacs thing. Still, if I can have the editor word wrap and use that, I’d feel a lot better about not having the problem of line breaks being interpreted by Github as new paragraphs.

**Comment by Pit-Storm** [comment]
Hi,

as we can see, Everyone has a diffrent configuration. So for me it make the most sense to let the default config, when screenreader mode is on, that line wrapping is disabled. But the limitation to not being able to turn it on again is really restrictive.

So please, just strip out the restriction to turn it on again 🙏

**Comment by falcon03** [comment]
As a blind user, I just wanted to advocate for letting us enable word-wrapping in VS-Code. If using it to write LaTeX code, for example, that feature is really convenient as single lines may become very long, and pretty unmanageable. When arrowing down to the next "logical" line, though, only the content of that "logical line" should be read, otherwise the feature would not be usable at all.

**Comment by falcon03** [comment]
Is there anything we can do as users to push this forward?

**Comment by mrgrieves4** [comment]
I am a visually impaired user and am losing my sight.  I use VoiceOver on the mac occasionally when my eyes are unable to focus on what I'm doing (or just for practice).

I often get prompted about using a screen reader.  Quite often I'm not currently at that particular moment in time, but will do from time to time.  So I'm never sure how to answer that question.

I find the inability to have wordwrap and voiceover in vscode a real pain, so I hope that a solution can be found at some point.  (But I appreciate how well vscode otherwise works with voiceover.)

**Comment by isidorn** [comment]
@mrgrieves4 thanks for your feedback. If you do not want to be prompted you can always explicitly set the 	`"editor.accessibilitySupport": "on"` in your settings.

As for word wrap not working there is a technical limitation why we are not doing this that I tried to explain here https://github.com/microsoft/vscode/issues/95428#issuecomment-614564888

**Comment by Pit-Storm** [comment]
Could it be a solution to talk to the team of NVDA to solve this in a sustainable way?

Or another solution could be insert a `\n` in a long line, that they are read correctly and in addition giving a shortcut that reads out "Line xxx part x of y."? So that you have the ability to get the information which line it is and how many wraps it have.

I am just thinking creative, because this line wrapping for a lot of people can be vary helpfull when writing e.g. markdown or latex code. And not be able to turn it on with read correctly by screenreaders for me is not very satisfying.

Why is it a problem to let the default behaviour like it is now and let the user decide to enable word wrap (with inserted `\n`)? Can you give us an specific answer to that question please, @isidorn ?

Thanks for your work and your awareness for that kind of topics :-)

**Comment by prater-devin** [comment]
I just use extensions to work around no wrapping in VS Code. If they won't
do it for us, we find a way. I believe the extension I use is called
"rewrap". It is because of this extension that I was able to do my work of
creating courses, in Markdown format, on Windows using VS Code. The fact
that an extension does what the VS Code team "cannot" do is interesting.
Thankfully, Linux, Emacs, and Emacspeak are a much more delightful
combination than VS Code probably ever will be. But for those that need
Windows and VS Code, and need to write in textual formats, the Rewrap
extension is there. Also the Grammarly extension is really great too.

Devin Prater

Technical Assistant


256 761-3423

[image: Alabama Institute for Deaf and Blind]

P.O. Box 698 | 1105 Fort Lashley Ave.

Talladega, AL 35161

www.aidb.org
[image: Facebook] Discover how we’re limitless.
<https://www.facebook.com/AlabamaInstituteforDeafandBlind?ref=notif&notif_t=fbpage_fan_invite>

**Comment by isidorn** [comment]
@Pit-Storm https://github.com/microsoft/vscode/issues/27766 this explains the limitation if we turn on word wrap with a screen reader.
Though I do agree we should invest in improving this experience and ideas are welcome.

**Comment by falcon03** [comment]
@isidorn sorry, but I am not understanding the technical limitation. Could you explain it better please?

The idea is that whenever word wrap is enabled, text navigation should work like those lines where actually different lines. So, for example, pressing arrow down should no longer scroll across "real" lines, but across "wrapped" lines. As a screen reader user, you can distinguish in between "real" linebreaks and the ones caused by wrapping because if you browse lines character by character you will encounter the newline character only if it is a "real" linebreak. Other editors behave this way, so this is something screen reader users are already familiar with.

Please let me know if I can help to push this issue forward... It's becoming more and more important for me, as VS-Code is becoming my IDE, but also the go-to program for writing LaTeX documents, markdown pages, and files in other formats that require using long paragraphs. Not having this feature makes working with such documents really painful, especially when you are reviewing them, fixing typos, etc.

**Comment by isidorn** [comment]
Thanks for the ping. We can enable word wrap when screen reader is attached but imagine there is a long sentence and it wraps into three lines A, B and C. When the user would navigate using up / down to B screen reader would read the whole line A, B and C (instead of just B). Same when the user would navigate to C or A the whole A, B, C would get read. So there is no way for us to let the screen reader know to only read the subset of the line, and not the whole thing.

@joanmarie is there some way to let the screen reader know it is a on wrapped line. We are using a `textArea` and I believe there is no API we could use for this. My guess is that other editors do not use a `textArea` and that gives them more flexibility.

I am open for ideas here. A workaround we could do is to produce a sound when the user would navigate to B or C so that when the screen reader would start reading A, B and C the user would know that she is on the wrapped part of the line hopefully.

**Comment by joanmarie** [comment]
@isidorn: I can triple-check, but a `textarea` should have the smarts to report (via accessibility APIs) the line contents at a specified offset. In which case, the question is: What is special about your `textarea` that prevents that?

**Comment by isidorn** [comment]
@joanmarie thanks! So everything works nicely when the actual word wrapping is left to the native `textArea`. The `textArea` client can not really control where the wrap will happen - it is up to the browser.
And this is where the challenge is for us: the editor decides where it is best to wrap based on font / view port / settings and other factors. And this is not the same as how our hidden `textarea` would wrap (it also does not have the same dimensions).
I do not think there is adding that can be fixed on the Chromium / screen reader side though, thanks for asking!

@alexdima @hediet do you have ideas how we could align the editor wrapping approach with the native web approach. Could we somehow achieve that the `textarea` wraps the text at the same place where the editor does?

**Comment by MarcoZehe** [comment]
So, in other words, the hidden textarea is so wide that it simply ignores the width of the actually displayed text editor on screen. Could the hidden textarea get a cols attribute set that corresponds to the character number the editor would be showing? That way, it would wrap the lines equally, because it gets told that the maximum number of characters per line should not be exceded.

**Comment by alexdima** [comment]
Good idea! To build on top of it, maybe instead of trying to get the `<textarea>` to match the editor's wrapping points, maybe we could turn it around and create a `<textarea>` with a width equal to the editor's, place the text inside of it, and then deduce the wrapping points from it. Then we could use those wrapping points as the editor's wrapping points. There are some wrinkles, for example we render tabs differently than `<textarea>`s do, and it will be tricky to keep a `<textarea>` that's so wide hidden visually, but this is something very much worth exploring!

**Comment by leonardder** [comment]
Note that https://github.com/microsoft/vscode/issues/97154 is related.

**Comment by isidorn** [comment]
@alexdima yeah that sounds promising! So let's put this on the plan as an exploration item when we find time.
@MarcoZehe currently the hidden `textarea` has wrapping disabled via css attributes. And it has a small size not comparable to the actual editor.

**Comment by ahicks92** [comment]
#141635 was closed in favor of this w.r.t. being able to manually format lines by hand for, e.g., C code.  I'm not sure how the two are related?  Should I be opening a separate issue?  Having lines wrap from the screen reader perspective isn't the same as being able to insert hard line breaks at the appropriate places for a style guide, or put another way "text is past ruler" != "editor is wrapping text".

**Comment by isidorn** [comment]
@ahicks92 please open a new issue explaining your use case precisely. Thanks! The other issue was closed because it was too general.

**Comment by byehack** [comment]
just wrap current focused line is just will be option that if it available i will enable it.

**Comment by isidorn** [comment]
🎉  👏  🚀
